Unit Testing, CLI Functionality, New Database, Beginning Project Expansion

CalciumKing opened the pull request:
>This is mostly a copy-paste from PR#4 with some slight changes.

<!--- Provide a general summary of your changes in the Title above -->
Fourth draft of the common stock inventory management project. Added general unit testing for controllers, specialized unit testing for `.py` files in `utils/`, CLI functionality and testing. All unit tests pass and app continues to function as intended. Project has been expanded to accommodate for "preventative maintenance", these expansions include: SQL Database, `ItemMaintenance` data class in `maintenance_data.py`, and enum class `Hutches` in `enums.py`. Project expansion so far has no functional changes, only preparations. All docstrings, requirement/setup files, and `README.md` are updated and up to date. New smart notifications feature notifies stock managers upon item total stock of 0. Also some minor bug fixes.

## Description
<!--- Describe your changes in detail -->
- Documentation:
   - Updated docstrings throughout project
   - Updated `README.md` with new features, usages, and tutorials
- Print statements and logging:
   - All print statements have been removed and replaced with more robust logging system:
   - `.info_log()`, `.warning_log()` and `.error_log()` have been removed
   - Logging method references have been replaced with `.info()`, `.warning()` or `.error()`, methods accompanied by `logging.getLogger()` used in several files where a logger reference is necessary.
- SQL:
   - Created a MySQL database per Joshua's request
   - Integrated MySQL database into the current system
   - Created some database dump backups
- Dual database system and database synchronization:
   - The system now uses a dual database system, MySQL and Google Sheets.
   - Upon app startup or on run of `python -m stock_manager sync`, the SQL database is updated to the state of the Google Sheet
   - This is to ensure that if someone edits the Google Sheet while the app is not running, it is then updated accordingly to ensure real time data tracking.
- Created a CLI system from scratch.
- Created a smart notification system that notifies stock managers via email upon item change if the total updated stock of the item is 0
- Improved some dunder processes in `item.py`'s Item data class to be more flexible and better handle processes
- Unit testing:
   - All controllers have general unit testing like changing pages, table functionality, etc.
   - All `utils/*.py` files have advanced unit testing, testing every method inside each of the scripts
   - CLI also has advanced unit testing which tests all available commands, testing commands multiple times for its short and long form if they have one
- Updated `.yaml`/`.yml` files to better use the GitHub pcdshub pipeline. Proper error handling in `__main__.py` and entry point for `cli.py`. Improvements in `item.py` allow comparing of `""` and `None`.


<details>
<summary>CLI Command Details</summary>

## CLI Top-Level Flags

| Command | Description |
|--------|--------|
| `-h`/`--help` | show a help message and exits |
| `-r`/`--run` | Runs Application GUI, Same As Running `python -m stock_manager` |
| `-v`/`--version` | Prints stock_manager's git version |
| `-t`/`--tree` | Prints stock_manager's commands, subcommands, and positional arguments in tree layout |

## CLI Top-Level Commands

| Header | Header |
|--------|--------|
| `export` | Exports All Item Data To A Specified File Type And Location (Used With `Path` And `Extension` Arguments) |
| `qr` | Generates A QR Code Of A Specified Item And Stores It As A `.png` In A Specified Location (Used With `Path` And `Part Number` Arguments) |
| `sync` | Synchronize Both Databases, Setting The Values Of The SQL Database To That Of The Google Sheet (Only Run If A Database Is Changed Externally) |
| `test` | Runs All Tests Located In `stock_manager/tests/` Using PyTest |
| `items` | Commands Related To Item Management (e.g., `list`, `add`, `search`, `remove`, `edit`) |
| `users` | Commands Related To User Management (e.g., `list`, `add`, `search`, `remove`) |

## Item CLI Subcommands/Flags

| Command/Flag | Description |
|--------|--------|
| `-l`/`--list` | Lists All Items In The Database |
| `search` | Searches For A Specified Value In All Fields In All Items (Used With `Part Number` Argument) |
| `add` | Add An Item With Specified Values To Both Databases (Used With Sequence Of 10 Item Detail Arguments) |
| `remove` | Remove A Specified Item From Both Databases (Used With `Part Number` Argument) |
| `edit` | Edit A Specified Item With At Least One Field Name And New Value (Used With `Part Number` Argument And Up To 9 Field Arguments) |

## User CLI Subcommands/Flags

| Command/Flag | Description |
|--------|--------|
| `-l`/`--list` | Lists All Users In The Database |
| `search` | Searches For A Specified Username In All Usernames (Used With `Username` Argument |
| `remove` | Add A User With Specified Username To Both Databases (Used With `Username` Argument |
| `remove` | Remove A Specified User From Both Databases (Used With `Username` Argument) |
</details>


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This PR updates the previous one with more updated documentation, a CLI, a second database, improved error messages, a more robust logging/printing system, bug fixes, and greater potential to be expanded in the next few weeks. Incorporating a CLI into the project has been in the project requirements since the very beginning as detailed in the project writeup. Various methods and functions across the program that once returned `None` now return a `bool`: `True` indicating that the method has finished successfully and no issues have occurred, `False` otherwise, this is important because it helps with unit tests and with general error checking and bug testing. Smart notification system was also introduced which was requested by Jasmine, a system that notifies her or the current stock manager when an item reaches a total stock of 0. Added unit testing which is also an industry standard and been part of the project requirements since the beginning, this is very important because it allows myself or other programmers in the future to test if their changes break any existing code.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Basic bug testing by myself has happened, unit testing has been created and used copiously. All tests pass when run and return the expected results (often `True`). As of the most recent push there are no known bugs. Unit testing will be continuously implemented as new features are created with the upcoming project expansion or further CLI development.

## Where Has This Been Documented?
<!--  Include where the changes made have been documented. -->
<!--  This can simply be  a comment in the code or updating a docstring -->
Docstrings have been updated in several modules, classes, methods, and functions - with accurate and completed information. Individual changes made have been documented in commit messages for each commit. Primary documentation for overall changes and improvements is this PR.

<details><summary>Screenshots</summary>
<img width="1489" height="384" alt="image" src="https://github.com/user-attachments/assets/77aeeabe-d7ac-4546-8df8-e99c1ec62cb6" />
<img width="321" height="43" alt="image" src="https://github.com/user-attachments/assets/0cd3de5c-b3ce-4670-91d9-d7a0648df6da" />
<img width="646" height="467" alt="image" src="https://github.com/user-attachments/assets/578faced-2276-4f67-bc92-32bfe99095e3" />
<img width="1121" height="266" alt="image" src="https://github.com/user-attachments/assets/469db8bd-e485-41c2-a5ef-d587daca922c" />
<img width="1682" height="304" alt="image" src="https://github.com/user-attachments/assets/22cba4dc-7946-4571-a634-01f8446af054" />
<img width="681" height="906" alt="image" src="https://github.com/user-attachments/assets/028968f0-69bb-4b26-9014-9b7ceecf2c23" />
<img width="1374" height="906" alt="image" src="https://github.com/user-attachments/assets/a76f3632-1638-4171-8212-0ce5102934a8" />
</details>

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [ ] Test suite passes on GitHub Actions
- [ ] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

CalciumKing committed:
Unit Tests and Backend Return Type Improvements

- Created tests folder and put unit testing python files and some assets used during testing inside it.

- Created test_utils.py to test all classes and most methods in the utils folder.

- Created test_controllers.py to test all controllers generally, test scanners generally, and began to test the add item page controller.

- In several files: added return types to make unit testing easier but also to return True if everything finished successfully and False if there was an issue of any kind. Also updated docstrings accordingly.

- Updated meta.yaml, requirements.txt, and README.md with new project (testing) requirements e.g. pytest and pytest-qt

CalciumKing committed:
Replaced second .gitignore with .gitkeep

- Files that end in _export are used in both testing and exporting, so I just added it to the main gitignore and replaced the second .gitignore with a .gitkeep

- This was mainly organizational, no application changes.

CalciumKing committed:
SQL Database Integration

- Added an SQL database to also be used in the project, acting as a more secure and reliable database, also including dumps in the new dumps folder for backups.

- In database.py: added mySQL code, updated outdated or incorrect error messages and code. Commented out google sheets API code like getting all data, getting all users, etc. since SQL is better but the google sheet API could still be used in the future. Updated .update_database() to be more efficient and work better with SQL, this involved getting rid of 'else' error messages for modifying items that did not exist in the database, these should be covered by the already existing exception catches.

- Updated import requirements in various files to reflect the new mySQL integration.

- Updated README.md file structure.

CalciumKing committed:
Database Notifications, Email Notifications, Notification Testing

- In controllers: added notification triggers when an item is created, edited, or scanned (subtracted) with a total stock of 0. A message is created, an email is sent, and the item name is added to its own page in the database.

- In test_utils.py: added database notification testing and email send testing.

- In scripts: created emails.py to test basic sending of emails.

- In utils: created notifications.py to put any functions that send notifications such as emails. Might rename this file in the future because it is unlikely that anything else will go inside it.

CalciumKing committed:
Beginning CLI Integration

- Created cli.py to build all the CLI commands and run extra code if necessary.

- Created testing/.../cli.py to test basics of CLI.

- Available Commands Include:
  -r/--run
  -v/--version
  -t/--test
  -li/--list-items
  -lu/--list-users
  -e/--export is in development

- In __main__.py: changed QApplication Arguments from [] to sys.argv to allow commands. Built commands before app is started so the user can run commands without starting the application.

- In setup.py: added an entry point to allow console scripts from __main__.py, formatting.

CalciumKing committed:
Formatting and Touch-Ups, No Functionality Changes

- Went through some files and formatted, replaced "" with '', removed unnececary type annotation, added small blocks for testing,

CalciumKing committed:
Synchronize Databases, Google Sheet Dominant Database, Fixed Database Loophole, Big Database Changes

- Fixed the issue of one database being more important than the other by making the Google sheet the dominant database.

- If changes are made to the Google sheet while application is not running, changes will be added/edited/deleted to/from the SQL database. SQL database is always synchronized to the Google Sheet database upon startup. Added very long .sync_databases() method in database.py to handle this.

- Added unit testing for new changes.

- Added instance of App in database.py as self._instance to get rid of NoneType errors for QMessageBoxes.

- Updated/Added docstrings where appropriate.

- Split .get_all_data() into .get_all_data_gs() and .get_all_data_sql()

- Split .get_all_users() into .get_all_users_gs() and .get_all_users_sql()

- In app.py: added self.db.sync_database() to ._async_load(), raises an exception to exit application if data cannot be synchronized. Gave DBUtils (self.db) a reference to self.

- In item.py: fixed __eq__ to work better when comparing '' and NoneType.

CalciumKing committed:
- In __main__.py: fixed error where .cli was causing app to crash, did not see this issue before because of strict CLI testing.

- In file_exports.py: renamed self.instance and self.logger to self._instance and self._logger since those variables are never referenced publicly.

- Added docstrings to all enums in enums.py

- Removed 'Stock Status' header from manual header setting since it already exists in GS database.

- Type annotation and spacing in test_utils.py.

CalciumKing committed:
Utils Changes, CLI Argument Creation

- Removed instance of App (instance) from both database.py and export.py, instead gave .sv_export() a new parameter all_items to export those items specifically.

- Moved .create_all_items() from app.py to database.py.

- Made .sv_export() not async and updated testing, removed failure testing for export because of Typing.Literal parameter.

- In cli.py: fixed some errors, updated positional arguments, -e/--export now works as intended, added functional -qr/--create-qr command.

CalciumKing committed:
Updated All Print Statements To Use .print() *args Functionality And Symbols

- Print statements use one of the following symbol prefixes to better style the message:

    [*]: Status
    [+]: Success
    [x]: Error
    [!]: Warning

- Reordered certain print statements to use , instead of f''

CalciumKing committed:
Test Fix, New Create Item Test

- Fixed issue of tests failing in test_utils.py.

- Added new test to test create_items database.py feature.

CalciumKing committed:
CLI Commands, Some Database Changes

- Fixed a line that was throwing an error with elon's jira api.

- In database.py: Renamed update_database to update_items_database, _update_sql to _update_items_sql, _update_gs, to _update_items_gs, and changed all occurrences. Changed some message strings, created _update_users_sql, _update_users_gs, and find_item. All are helper methods that interact with at least one of the databases.

- In cli.py: changed multiple optional positional arguments to a single optional positional argument array, replaced all older usages with this new system. With new system added parameter checking for all commands. Changed _run_list_items and _run_list_users to allow searching. Added functional commands below:

   -si/--search-items
   -su/--search-users
   -ai/--add-item
   -au/--add-user
   -ri/--remove-item
   -ru/--remove-user
   -s/--sync-databases

CalciumKing committed:
CLI Backend Changes

- Changed CLI building to be more efficient and less confusing for anyone viewing code.

- Moved all if/elif code to their own _run_... functions.

- Added .validate_args() to check if the number of arguments is correct. Checking if arguments are valid using .validate_args() and possibly returning False is the first line of all _run_... functions.

- Moved all commands to a command dictionary mapping instead of a long if-elif-else statement.

- Moved all argument builds to an array of tuples to hold the short keyword, long keyword, help description. This system only uses one parser.add_argument() command since they are all using action='store_true'.

- Overall, efficiency and line count changes, no frontend or user changes at all.

CalciumKing committed:
Added/Updated Docstrings

- Updated various docstrings around project

- Added docstrings in database.py, enums.py, notifications.py, and cli.py

- Renamed validate_args to _valid_args because it is only used locally.

CalciumKing committed:
Renamed tests/assets to tests/exports, Test Fixes

- Renamed assets to exports to be more accurate and change project compiler settings. Updated references accordingly.

- Fixed test_qr_checking issue in test_controllers.py.

CalciumKing committed:
cli.py Backend Changes, Started test_cli.py

- Renaming changes in cli.py, no frontend changes.

- Created test_cli.py to test cli.py commands. In development. All tests currently work.

CalciumKing committed:
CLI testing, New CLI Command.

- in test_cli.py: added more unit testing, all commands now have proper testing. Moved no_args_command() to pytest parameters.

- In cli.py: added -ei/--edit command to edit an item after specifying its name and at least one field to change, updated print statements, updated error checking in various methods.

CalciumKing committed:
item.py Docstrings, item.py __setitem__

- In item.py: updated either outdated, non-informational, or otherwise unfinished docstrings. Added __setitem__() to allow dictionary-like index setting (item['field_name'] = ...).

- Updated return types in two functions.

CalciumKing committed:
New Project Expansion Additions

- Added new enums for the various Hutches at SLAC, used only in the new maintenance_data.py dataclass.

- Created first draft of an ItemMaintenance dataclass to be used in the new preventative maintenance expansion.

CalciumKing committed:
Update README.md

CalciumKing committed:
Merge branch 'pcdshub:master' into master

CalciumKing committed:
Merge branch 'master' of https://github.com/CalciumKing/SLAC-LCLS-Stock-Management

CalciumKing committed:
Merge branch 'master' of https://github.com/pcdshub/stock-management

CalciumKing committed:
Merge branch 'master' of https://github.com/CalciumKing/SLAC-LCLS-Stock-Management

CalciumKing committed:
Requested #4 Changes - Commit 1

- Modified logger.py to also print to the console, removing the need for print statements and console severity symbols.

- In logger.py: Updated docstrings, deleted static methods, as loggings are called using logging.getLogger() and .info, .warning. or .error. Added a handler to the basicConfig call to cause logs to be printed to the console as well as to app.log.

- Removed print statements in all modules not located in scripts or tests. Updated or added logs as necessary.

- In test_utils.py: updated logging tests accordingly.

CalciumKing committed:
Requested #4 Changes - Commit 2

- Used prettytable instead of custom ASCII table for cli.py.

- Fixed some logic and docstrings/print statements.

CalciumKing committed:
Requested #4 Changes - Commit 3

- In __main__.py: uncommented try-except block, turned print statement into logging call, updated main() docstring.

CalciumKing committed:
Requested #4 Changes - Commit 4

- In scripts/: deleted blank __init__.py, created a sub README.md for scripts/. README.md contains information explaining that everything inside scripts/ is for development only and should be treated as experimental.

- In txt_export.txt: added an explanation as to why the blank file is necessary and what it does for unit testing.

CalciumKing committed:
Requested #4 Changes - Commit 5

- Added use of caplog to tests in test_cli.py so the unit tests are actually testing if the command finished properly and that they finish without errors and not just finishing without errors.

CalciumKing committed:
Requested #4 Changes - Commit 6

- Changed cli.py to build arguments as parsers and not all top level positional arguments. I.e., ...stock_manager items edit sample_item... and not ...stock_manager --edit sample_item...

- Updated docstrings in cli.py.

- Removed entry point and concept of always returning False, cli argument functions now return True if finished successfully and False otherwise.

- Remade entire argument building system to be more straightforward and tailored to this project instead of general top-level arguments.

- Got regions working and added them to organize cli.py better.

- Removed _valid_args() everywhere because argparse already validates argument count.

CalciumKing committed:
Requested #4 Changes - Commit 7

- Updated test_cli.py according to the new changes in CLI functionality.

- Actually testing if command is finished successfully and not just no errors.

- Added better assert statements that show the contents of captured.out if success message is not included.

CalciumKing committed:
Create notifications.py

CalciumKing committed:
Misc Touch-ups and Fixes

- Fixed keyboard shortcut bug with GUI, keyboard keys would be 1 off from actual desired page. Fixed error message bug on exports page, when no file extension is selected, incorrect error message would display.

- Added help text to extension argument used with export command.

- Changed nav bar button appearance in GUI. Made scanned items list look bigger and more readable.

- Added spacing to GitHub templates.

CalciumKing committed:
GitHub Pipeline Testing

- Learning how to use the GitHub pipeline to pass pcdshub tests.

CalciumKing committed:
Unit Test Fixes

- Fixed email notification error

- Removed some test skipping by making a new fixture for controllers with tables.

tangkong commented on .github/workflows/standard.yml:
```diff
@@ -1,25 +1,22 @@
 name: PCDS Standard Testing
 
+permissions:
+  id-token: write
+  contents: read
+
```
Why did you need to adjust the default permissions here?

CalciumKing replied:
I'm getting errors with `standard.yaml` that prevent `pcdshub` tests from running, the errors relate to the file's or the repo's permissions-I have no experience with `.yaml`/`.yml` files and I found that code bit online and thought it related to my issue but I could be wrong. I can delete it in the next commit if you don't think it does anything or is unnecessary.

tangkong replied:
I was a bit confused here as to why the tests weren't running here in the first place. I mostly point this out because none of our other CI setups need these settings, and we shouldn't be granting any write permissions to begin with.

These yaml files are just configurations for the github actions runner, it's less a language to learn and more of learning how to configure github actions.

tangkong replied:
It looks like the errors relate to us requesting write permissions, where the shared job we're using doesn't allow it

> The nested job 'deploy' is requesting 'contents: write, pages: write', but is only allowed 'contents: read, pages: none'.

tangkong commented:
Before I dive too deep here, I think some bullet points in your pull request description would be really help.

tangkong commented on conda-recipe/meta.yaml:
```diff
@@ -13,16 +13,11 @@ requirements:
-    - qrcode
-    - qtawesome
+    - pip:
+        - -r requirements.txt
```
In general I think it's better to keep the dependencies listed explicitly, rather than rely on the pip requirements list.  While it's a bit of duplication, conda has its own dependency resolution system and using pip circumvents that.  

This gets into a bit of the details of packaging, but it's best to not mix conda and pip unless it's absolutely necessary.

CalciumKing replied:
I can add all the requirements back in the next commit.

tangkong commented on stock_manager/tests/test_utils.py:
```diff
@@ -0,0 +1,107 @@
+        return DBUtils()
+    
+    def test_sync_databases(self, database):
+        assert database.sync_databases()
```
While this pattern (returning True if a method thinks it was successful) might seem convenient, I think it's actually not a very effective test.

What we're trying to accomplish with unit testing is verifying that our code is doing what we expect it to do.  Normally this manifests as performing some action, then checking (asserting) that action had the expected effect.  

In this scheme, I could delete the entire body of the method except for returning True, and the tests would still pass.  This is a bit of hyperbole, but any number of changes could happen to these methods that break the functionality of the app and the tests wouldn't help us at all.

I think these tests could use asserts that inspect the objects being modified.  For example here you could check the state of the database before and after the sync.

CalciumKing replied:
I see what you mean, however I feel it's important to mention that some test methods like `test_fetch...()`, `test_create_items()`, `test_fetch_headers`, etc. are simple, one line test functions because they don't return a bool, they return a real value that is meant to be used somewhere else in the project, just having `assert function_call()` is asserting that the value is returned instead of an error or `None`.
I will go through and replace the test functions that just test if `True`/`False`, but I feel it could be useful in the future to leave the return types of the functions that return a `bool` based on the success of the function.

tangkong replied:
Totally understand, there's a time and place for most patterns.

I'm not necessarily saying that returning a bool based on perceived success is a bad pattern (though I favor assuming success unless an exception is thrown).  I was more suggesting that the test shouldn't rely on only the fact that the value is "truthy".  We can do some more in-depth inspection of the effects.

tangkong commented on stock_manager/tests/test_utils.py:
```diff
@@ -0,0 +1,107 @@
+
+
+class TestLogger:
+    Logger()
```
Broadly, is this just testing the python built-in logger?  If so I'd argue that's not something we need to do.

CalciumKing replied:
No, `Logger()` is the first line of the class so when any of the tests are run, a project-wide logger is initialized. This allows for the methods in `TestLogger` to reference the correct logger and not create their own with `logging.getLogger()`.

CalciumKing replied:
I revisited this code and modified it to work with `caplog` so it can actually test if the message is being logged instead of just checking if it returns `None` and no errors are thrown. This means that the call to `Logger()` is no longer necessary and can be removed in the next commit.

tangkong commented on stock_manager/tests/test_utils.py:
```diff
@@ -0,0 +1,107 @@
+        assert database.update_items_database(
+                update_type, [Item(
+                        *['None'] * 3,
+                        *[0] * 6,
```
I originally wanted to comment that this was unnecessarily code-golf-y.  It's more readable to actually list the arguments, even if it's repetitive.

Now I'm suspicious that an `Item` with almost entirely null values is a valid addition to the database.  Should `Item`s have some minimal set of required information?

CalciumKing replied:
I will change the code to list all the values without the list generation and expansion in the next commit, also the app does not allow the setting of `None` values, ensuring that all new items and updated old ones are all filled out properly however in `item.py`, each field has a `| None` alternative, this is because there are entries in the database that are incomplete and have empty values, ideally in the future when this app is used consistently, there should be no `None` values in the database and all the `| None` alternatives values can be removed. I believe the confusion lies in the fact that those aren't actually `None` values, those are `"None"` values, as in strings. I also changed the `"None"` to `"Test"` so it is less confusing.

tangkong commented on stock_manager/tests/test_utils.py:
```diff
@@ -0,0 +1,107 @@
+                update_type, [Item(
+                        *['None'] * 3,
+                        *[0] * 6,
+                        stock_manager.StockStatus.OUT_OF_STOCK.value
```
In the dataclass here you should just hint it as the `StockStatus` enum, and pass the enum itself around.  rather than its value.  Doing it this way sort of defeats the purpose of setting up the enum

tangkong commented on conda-recipe/meta.yaml:
```diff
@@ -13,16 +13,11 @@ requirements:
-    - qasync
-    - qrcode
-    - qtawesome
+    - pip:
+        - -r requirements.txt
```
Despite the repetition, it's generally better to keep the requirements explicitly listed here.  Conda has its own dependency resolution system, and using pip circumvents that.   The rule of thumb is to never mix conda and pip installs if you can help it.  

This is a packaging nitpick that you probably shouldn't have to care about, but figured it's worth a mention.

tangkong reviewed (commented):
Leaving a partial review, focusing mostly on the tests.  Lots of good work here, don't let my predominantly critical review comments discourage you

In general I'm not a fan of the patten where every method returns a bool indicating its success.  In the context of testing, it doesn't actually offer us any guarantees that the business logic is operating as expected.  In the context of the app, is every method expected to check the success status of the methods they call and raise?  

Open to differing opinions here, but I favor raising exceptions if something unexpected happens.

CalciumKing committed:
Requested #5 Changes - Commit 1

- In meta.yaml: fixed conda import circumvention by removing pip. Added testing imports to test section.

- In test_utils.py: fixed "code-golf-y" test statement and changed text values from "None" to "Test" to remove NoneType confusion.

- In item.py: changed stock_status field from string (enum value) to just enum and updated its references with grabbing the enum's value.

In standard.yml: removed unnecessary permissions block.

CalciumKing committed:
CLI Command Tree Command

- In cli.py: added command that allows users to view commands as tree format to better understand the order of the commands rather than repetitively typing an argument and -h/--help until they run the correct command.

CalciumKing committed:
Update meta.yaml

CalciumKing committed:
Reverted abstract.py Back To Original State, Modified item.py to accommodate for Enums instead of references to item.py dunder methods.

CalciumKing committed:
Added tree command testing similar to --help command testing

CalciumKing committed:
Huge Unit Testing Changes

- In conftest.py: added testing item, username, and notification to be used in various tests around the tests/ folder.

- In test_cli.py: replaced certain string usages with references to test item/username/notif from conftest.py.

- In test_controllers.py: improved testing on already existing tests, added testing for already existing classes and methods, and added lots of new testing for controllers that didn't have unit testing of any kind.

- In test_utils.py: optimized testing and made it test better instead of just testing if something was/wasn't true or threw errors or not.

- In controllers/: changed some methods to have return types to help with unit testing.

- In constants.py, database.py, item.py: fixed some bugs that were causing problems or should have already been implemented.

- All tests should actually test the feature the method/function name describes.

CalciumKing committed:
Misc Formatting/Shortening

- In tests/: shortened decorators by importing mark/fixture/asyncio from pytest and referenced that instead of the full call.

- Fixed formatting of some code, added or removed whitespace for better organization.

- Changed some QMessageBox debug levels to be questions instead of what it was before to fit the prompt better.

- Changed some abstract method content in abstract.py from pass to ... to show that it is implemented elsewhere and not just placeholder code.

CalciumKing committed:
.yaml/.yml Changes

CalciumKing committed:
Update meta.yaml

CalciumKing committed:
Returned meta.yaml back to original state to see if tests run successfully at all

CalciumKing committed:
Update meta.yaml

CalciumKing committed:
Attempting to remove pytest-qt from requirements so GitHub actions will pass. Also removed its use from test_controllers.py since qtbot is a part of pytest and QtBot was only being used for type annotation.

CalciumKing committed:
Fixing Python Import Problems

- In __init__.py: changed __all__ to work properly and be list[str] and not list[str | list[str]].

- Updated all import statements and local references accordingly.

CalciumKing committed:
Flake8 Whitespace And Line Length Changes

- Changing main file lengths to be <79 characters to pass Flake8 .pre-commit-config.yaml/GitHub test

- In .pre-commit-config.yaml: removing unnecessary checks that aren't used in this project.

- Excluding scripts/, versioneer.py, and _version.py from checks.

- Removed elon_jira_api.py: JIRA is not being used, jira_api.py stays for possible future example, but for now, elon's jira file can go

- In README.md: removed trailing spaces for <br> tag to put text on next line.

CalciumKing committed:
Fixed .pre-commit-config.yaml To Exclude versioneer, _version, and scripts/ Instead Of Just Excluding For Flake8

CalciumKing committed:
Update meta.yaml

CalciumKing committed:
Updated Building Requirements For GitHub Tests, Modified setup.py To Hopefully Fix Issue

CalciumKing committed:
Updated Building Requirements For GitHub Tests, Modified setup.py To Hopefully Fix Issue

CalciumKing committed:
Updated meta.yaml

- Added some code that will hopefully fix the issue of the GitHub actions tests not being able to find stock_manager.

- Also updated the python version to work with the rest of the system, setup.py, and what is already specified in the README.md.

CalciumKing committed:
Updating meta.yaml

- Updating the source_files line to accurately find the tests folder instead of just searching the root directory for a tests folder that doesn't exist.

CalciumKing committed:
Updated 3 Files To Use Python 3.9 Instead Of Python 3.10 Because Of Conflicting pcdshub Python Versions

CalciumKing committed:
Removed all instances of "override", Including All @override Decorators Due To Inexistence In Python 3.9

CalciumKing committed:
Added QtBot Reference back To test_controllers.py Because PyTest-Qt Should Work Now

- Also added pytest-qt to requirements.txt

CalciumKing committed:
Fixed __iter__ Issue In item.py

- Changed return type of Item.__iter__ from Generator[...] to Iterator[...] to be more accurate to iterator function.

- Updated __iter__ docstring slightly.

CalciumKing committed:
GitHub Actions Bug Fixes

- In meta.yaml: Error Occurring With Testing Environment, Trying Custom Testing Command To See If It Works Better.

- In item.py: changed | types to typing.Optional[...] instead for better compatibility with python 3.9.

- In setup.py: added os path joining to get the current directory of setup.py instead of just referencing requirements.txt locally. No issue locally but GitHub actions testing environment has a different structure than local repo.

CalciumKing committed:
Including requirements.txt in MANIFEST.in so it is included in the project build.

CalciumKing committed:
| Operator To Union Changes, pip requirements.txt Change

- In requirements.txt: changed cv2 to opencv-python for actual python pip package.

- In item.py: changed long annotations that include | operator to use typing.Union[...] instead for compatibility with python <3.10

CalciumKing committed:
Update item.py

Replacing Union | with ,

CalciumKing committed:
Iterator Bug Fix, mysql pip Package Name Fix

CalciumKing committed:
Jasmine Push

CalciumKing committed:
Fixed Union Bugs, Cleaned Up Project Hierarchy And .gitignore, Updated README.md

CalciumKing committed:
Bug Fixes, Google Sheet Database Change, Minor Changes

- In abstract.py - Set Table Headers Instead Of Fetching Them From Database Because Of Extra Fields

- Changing Database From Custom Google Sheet To Actual ECS Common Stock Inventory Google Sheet

- Removed test_fetch_headers in test_utils.py and get_headers from database.py.

- In database.py: MySQL Database is now optional, meaning the program doesn't need it to continue and can just use the Google sheet instead. Checks have been put into place to prevent MySQL use if it doesn't exist. Updated error messages.

CalciumKing committed:
Bug Fixes And Minor Changes

- Unwanted Logging
- Close Event Crash
- Users Database Fetching Header As Username

- Missed SQL Check From Last Commit
- Line Length Fixes

CalciumKing committed:
Removed All Notification and Email Sending Functionality

- In database.py: removed notification functionality and all references.

- Removed notifications.py and all references.

- In controllers/: removed all notification setting and email sending functionality because google sheets has build in functions that do what I tried to do but better, implemented in google sheet.

- In test_utils.py: Removed all email and notification testing, fixed imports.

tangkong commented:
So the CI seems to be failing for variety of reasons.  I think a reasonable approach here is to merge this without tests passing, then go back and get the tests passing in a separate effort.

* Conda 3.9: build process fails to complete due to using syntax introduced in later versions of python
  * The proper thing to do here is specify in our requirements that the application is not compatible with lower versions of python.  In the short term we could probably ignore this test since we're moving toward
* Conda 3.12: An actual failure in the test suite.  
  * There's no debug output here since you're invoking pytest on your own from inside the package recipe.  You could rely on our workflow here, which calls pytest with a bunch of verbose debug output enabled ([ci-helpers link](https://github.com/pcdshub/pcds-ci-helpers/blob/34ff5c516726af13387713ca76ed6079b361098b/.github/workflows/python-conda-test.yml#L227-L235))
* Pip 3.9: same failure as the conda 3.9
* Pip 3.12: Actual fest suite failure
  * The traceback here seems suspiciously like a segfault.  it looks like the test is attempting to sync the two databases.   Have you set up test databases for the test suite to talk to inside of the CI runner?
* Pip 3.13: see above
* Documentation building: You don't have a docs folder
  * We're expecting a pretty specific docs directory structure, so we can have sphinx build html docs for us.  It's probably prudent to just ignore this test for the time being, but you can refer to [our cookiecutter](https://github.com/pcdshub/cookiecutter-pcds-python/tree/master/%7B%7B%20cookiecutter.folder_name%20%7D%7D/docs) if you really want.


For what it's worth, the reason we set these tests up is to make sure the application runs in a new environment, outside the bias of the developer's environment.  Getting the app running locally is a challenge, but getting it set up reproducibly is the next step in the challenge.  You're fighting the systems we spent a while building/standardizing, and I should have probably sat you down and explained what was going on.  (apologies)

CalciumKing committed:
Added New Database Dumps

CalciumKing committed:
Update README.md

CalciumKing commented:
I pushed the final commit for this week. All my tests work locally and I've gotten to a point where the tests are starting to run in the CI, which I will work on next week as well. This is the last major PR for my internship - possibly one more short PR next week if any testing changes are made. This PR has enough new changes and better testing that we can merge without issue.

ZLLentz commented:
I'm going to merge this so we can test from master branch next week!
